test/e2e_federation/util.go

```diff
@@ -172,33 +226,70 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
 		}
 		return false, err
 	})
+	return err
+}
+
+func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, serviceName string, orphanDependents *bool) {
+	if clientset == nil || len(namespace) == 0 || len(serviceName) == 0 {
+		Fail(fmt.Sprintf("Internal error: invalid parameters passed to deleteServiceOrFail: clientset: %v, namespace: %v, service: %v", clientset, namespace, serviceName))
+	}
+	framework.Logf("Deleting service %q in namespace %v", serviceName, namespace)
+	err := deleteService(clientset, namespace, serviceName, orphanDependents)
 	if err != nil {
 		framework.DescribeSvc(namespace)
-		framework.Failf("Error in deleting service %s: %v", serviceName, err)
+		framework.Failf("Error in deleting service %v in namespace %v err: %v", serviceName, namespace, err)
 	}
 }
 
+func getServiceShard(namespace string, serviceName string, cluster *fedframework.Cluster) (*v1.Service, error) {
+	name := cluster.Name
+	var cSvc *v1.Service
+	err := wait.PollImmediate(framework.Poll, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
+		var err error
+		cSvc, err = cluster.Clientset.Services(namespace).Get(serviceName, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			// Get failed with an error, try again.
+			framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v. Trying again in %s", serviceName, namespace, name, err, framework.Poll)
+			return false, nil
+		} else if errors.IsNotFound(err) {
+			cSvc = nil
+			By(fmt.Sprintf("Service %q in namespace %q in cluster %q not found", serviceName, namespace, name))
+			return true, err
+		}
+		By(fmt.Sprintf("Service %q in namespace %q in cluster %q found", serviceName, namespace, name))
+		return true, nil
+	})
+
+	return cSvc, err
+}
+
+func getServiceShardWithRetry(namespace string, serviceName string, cluster *fedframework.Cluster) (*v1.Service, error) {
+	name := cluster.Name
+	var cSvc *v1.Service
+	err := wait.PollImmediate(framework.Poll, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
+		var err error
+		cSvc, err = cluster.Clientset.Services(namespace).Get(serviceName, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			// Get failed with an error, try again.
+			framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v. Trying again in %s", serviceName, namespace, name, err, framework.Poll)
+			return false, nil
```

Comment:
it would be better to `return false, err` here, if any err other than notfound err. Since it may be non-recoverable error and there is no point in keep on retrying.